### ISSUE TITLE
Consumer is re-created after subscription is deleted

### ISSUE THREAD

**Issue by samuel-beniamin** [allegro/hermes#557]
After deleting a subscription and the consumer stopped as expected. The work load balance kicks in with a signal for a new assignment created for the already deleted subscription therefore a new consumer is started for new assignment. At least this is what I can figure from the logs [here](https://gist.github.com/samuel-beniamin/12819bab1434084f5fd73239d027b38a), I've also left some comments on the logs. What ends up in Zookeeper is that there is no node for the subscription yet there is an assignment node under the `/runtime`.

The used configuration was on my local machine with a single instance of the consumer:
* consumer.workload.algorithm=selective
* consumer.workload.rebalance.interval.seconds=30
* consumer.workload.consumers.per.subscription=1
* consumer.workload.max.subscriptions.per.consumer=200

**Comment by chemicL** [comment]
Hey, @samuel-beniamin.

Thanks for the report. I tried to reproduce the issue with no success.

Have you modified anything in the source code? From what I investigated, the situation happening in your logs should not be possible in the latest implementation of the selective workload algorithm. Namely events happening in the same moment in time:

`14:46:11 [INFO ] [p.a.t.h.i.z.c.HierarchicalCacheLevel] [] Got CHILD_REMOVED event for path /hermes/groups/test.topicowner/topics/deleatable-topic/subscriptions/test.topicowner-first
14:46:11 [INFO ] [p.a.t.h.i.z.c.HierarchicalCacheLevel] [] Got CHILD_REMOVED event for path /hermes/consumers-workload/primary/runtime/test.topicowner.deleatable-topic$test.topicowner-first/localhost$482245515483069223`

while only place where `CHILD_REMOVED event for path /hermes/consumers-workload/primary/runtime/test.topicowner.deleatable-topic$test.topicowner-first/localhost$482245515483069223` could have happened is in the `BalancingJob`, which has been triggered before at 14:45:48 and later 14:46:18.

Please let me know if there's any particular steps involved in reproducing the situation on your side, as it seems you got it covered ;-)

**Comment by samuel-beniamin** [comment]
Hey @chemicL,

Thanks for taking the time to have a look on the issue.

Yes there are some code changes in hermes/consumers e.g. adding more metrics or Gradle configuration, all trivial changes but nothing related to `BalancingJob` nor the selective workload algorithm or any logic change.

1. We are using hermes as a dependency, so our custom logic makes the call to hermes Java APIs like for example when deleting a subscription we do call `void removeSubscription(TopicName topicName, String subscriptionName);` from `SubscriptionRepository`.
2. I had the test running on four Kafka brokers and a single zookeeper node, as a single instance hermes consumers (as a dependency)
3. Topic config: Min-in-sync-replicas are 2, and replication factor is 3, on my local machine.

If you have any questions regards something specific, please let me know.

**Comment by chemicL** [comment]
With one consumer instance we can rule out any clock skews in the logs in such case.

As I noted, the only way the suspicious combination of logs could happen would be if the BalancingJob executed the assignment shuffling.

Namely, `SubscriptionAssignmentRegistry#dropAssignment` method would need to be called explicitly for the zookeeper node to be removed, as we can observe in the log.

Without seeing the code I won't be able to help much I'm afraid.

Are you by any chance able to reproduce this on vanilla hermes code from github?

**Comment by adamdubiel** [comment]
I think we need to add a little disclamier here: the things you are using are internal API, which can change pretty much at any point - not saying it will, but we won't think twice about change if we have the need to do it.

**Comment by samuel-beniamin** [comment]
I'm still able to reproduce with a vanilla hermes code `0.8.10-hotfix2`. 

After deleting the subscription there was still messages tried to be sent to the subscriber, although the subscriber was returning `401 Unauthorized` because of wrong basic credentials info in the subscription while it also had `retryOnClientError: true`.

I've provided [logs](https://gist.github.com/samuel-beniamin/e61f1a19ac1e9eaae29ecf81edf6c7f6) on Debug level showing that after deleting the subscription it was still trying to send messages to the subscriber.

Here is the state that ends up in zookeeper:
```bash
ls /hermes/groups/group/topics/my-topic/subscriptions
[]

ls /hermes/consumers-workload/primary/runtime
[]
```

**Comment by samuel-beniamin** [comment]
It persists until I restart the whole consumers app, I've already provided some logs on my comment. But here it is a bit detailed in debug level https://gist.github.com/samuel-beniamin/e61f1a19ac1e9eaae29ecf81edf6c7f6

**Comment by samuel-beniamin** [comment]
One update is that in this particular test the sending stopped after 20 mins, but for other cases we could notice that is could stay at least for few hours. One more thing that on another test we have found the assignment still exist in Zookeeper as well.

**Comment by samuel-beniamin** [comment]
Were you able to reproduce or find this as a bug ?

**Comment by samuel-beniamin** [comment]
After further investigation we found that in the `ConsumerMessageSender` the `retrySingleThreadExecutor` was never shutdown. After changing the shutdown method of `ConsumerMessageSender` to be 
```java
public void shutdown() {
  running = false;
  retrySingleThreadExecutor.shutdownNow();
}
```
That fixed the problem for us. I although wonder if there could be a better solution in your opinion.

**Comment by adamdubiel** [comment]
I see the same behavior on some thing i deleted on our prod, we will fix it in next version :)

**Comment by adamdubiel** [comment]
Fix for this issue is on the way https://github.com/allegro/hermes/pull/583 and is going to make it to `0.8.12`. Thanks for reporting & finding this, i think this caused some serious problems on our prod recently.

**Comment by adamdubiel** [comment]
Closed with #583
